bosond.cpp

```diff
@@ -99,125 +106,133 @@ int main(int argc, char** argv) {
+        headers << '\n';
+        auto header_str = headers.str();
+        if (sendall(sock, header_str.data(), header_str.length()) < 0) {
+            perror("HEADERS");
+            exit(1);
+        }
     }
 
     // Activate streaming
-    int type = bufferinfo.type;
-    if(ioctl(fd, VIDIOC_STREAMON, &type) < 0){
+    if (ioctl(fd, VIDIOC_STREAMON, &bufferinfo[0].type) < 0) {
         perror("VIDIOC_STREAMON");
         exit(1);
     }
 
-    system_clock::time_point t0 = system_clock::now();
+
+    steady_clock::time_point t0 = steady_clock::now();
+
+    // Indices of the buffer being filled (i_active) and the buffer
+    // that is full and is ready to send (i_full)
+    int i_active = 0;
+    int i_full = 1;
+
+    // Put the first buffer in the incoming queue.
+    if (ioctl(fd, VIDIOC_QBUF, &bufferinfo[i_active]) < 0) {
+        perror("VIDIOC_QBUF");
+        exit(1);
+    }
+
     int count = 0;
     for (;;) {
-        // Put the buffer in the incoming queue.
-        if(ioctl(fd, VIDIOC_QBUF, &bufferinfo) < 0){
-            perror("VIDIOC_QBUF");
+        // Wait for active buffer to be filled.
+        if (ioctl(fd, VIDIOC_DQBUF, &bufferinfo[i_active]) < 0) {
+            perror("VIDIOC_DQBUF");
             exit(1);
         }
 
-        // The buffer's waiting in the outgoing queue.
-        if(ioctl(fd, VIDIOC_DQBUF, &bufferinfo) < 0) {
+        std::swap(i_active, i_full);  // Switch buffer roles
+
+        // Put next buffer in the incoming queue so the video driver
+        // can start filling it.
+        if (ioctl(fd, VIDIOC_QBUF, &bufferinfo[i_active]) < 0) {
             perror("VIDIOC_QBUF");
             exit(1);
         }
 
+        // Send full buffer while other buffer is being filled.
+        if (sendFrames) {
+            if (sendall(sock, (const char *) buffer[i_full], bufferinfo[i_full].length) < 0) {
+                perror("SEND");
+                exit(1);
+            }
+        }
+
         if (printTimings) {
             count++;
-            if (count == 100) {
-                system_clock::time_point t1 = system_clock::now();
-                std::cout << "td = " << std::chrono::duration_cast<std::chrono::microseconds>(t1 - t0).count() << "[µs]" << std::endl;
+            if (count == 120) {
+                steady_clock::time_point t1 = steady_clock::now();
+                auto us = duration_cast<microseconds>(t1 - t0).count();
+                auto rate = count / ((float)us / 1e6);
+                std::cout << rate << "Hz" << std::endl;
                 t0 = t1;
                 count = 0;
             }
         }
-
-        if (sendall(sock, (const char *) buffer_start, bufferinfo.length) < 0) {
-            perror("SEND");
-            exit(1);
-        }
     }
 
-    // Deactivate streaming
-    if( ioctl(fd, VIDIOC_STREAMOFF, &type) < 0 ){
-        perror("VIDIOC_STREAMOFF");
-        exit(1);
-    };
-
-    close(fd);
     return 0;
```

Comment:
is deactivating the stream not needed?

Reply:
No it's not actually ever called but I thought I'd leave it there for completeness. We might choose to add a signal handler which tidies things up at some point.

I guess it's probably more confusing than helpful to have unused code sitting there so I'll remove it.